WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt

```diff
@@ -1,35 +1,47 @@
 import org.wordpress.android.ui.mysite.MySiteCardAndItemBuilderParams.PostCardBuilderParams
+import org.wordpress.android.ui.mysite.cards.post.mockdata.MockedPostsData.Post
+import org.wordpress.android.ui.utils.UiString.UiStringRes
 import org.wordpress.android.ui.utils.UiString.UiStringText
 import javax.inject.Inject
 
 class PostCardBuilder @Inject constructor() {
-    fun build(params: PostCardBuilderParams): List<PostCard> {
-        val cards = mutableListOf<PostCard>()
-        params.mockedPostsData?.posts?.draft?.map {
-            cards.add(
-                    PostCard(
-                            title = UiStringText(DRAFT_TITLE),
-                            postTitle = UiStringText(it.title ?: NO_TITLE)
-                    )
-            )
-        }
-        params.mockedPostsData?.posts?.scheduled?.map {
-            cards.add(
-                    PostCard(
-                            title = UiStringText(SCHEDULED_TITLE),
-                            postTitle = UiStringText(it.title ?: NO_TITLE)
-                    )
-            )
-        }
-        return cards
+    fun build(params: PostCardBuilderParams) = mutableListOf<PostCard>().apply {
+        val posts = params.mockedPostsData?.posts
+        posts?.draft?.takeIf { it.isNotEmpty() }?.let { add(it.createDraftPostsCard()) }
+        posts?.scheduled?.takeIf { it.isNotEmpty() }?.let { add(it.createScheduledPostsCard()) }
     }
 
-    companion object {
-        const val DRAFT_TITLE = "Draft Post"
-        const val SCHEDULED_TITLE = "Scheduled Post"
-        const val NO_TITLE = "No title"
+    private fun List<Post>.createDraftPostsCard() = PostCardDraftOrScheduled(
+            postCardType = PostCardType.DRAFT,
+            title = UiStringRes(R.string.my_site_post_card_draft_title),
+            postItems = mapToDraftOrScheduledPostItems(PostCardType.DRAFT)
+    )
+
+    private fun List<Post>.createScheduledPostsCard() = PostCardDraftOrScheduled(
+            postCardType = PostCardType.SCHEDULED,
+            title = UiStringRes(R.string.my_site_post_card_scheduled_title),
+            postItems = mapToDraftOrScheduledPostItems(PostCardType.SCHEDULED)
+    )
+
+    private fun List<Post>.mapToDraftOrScheduledPostItems(postCardType: PostCardType) = map { post ->
+        val excerpt = if (postCardType == PostCardType.SCHEDULED) {
+            "Today at 1:04 PM" // TODO: ashiagr - remove hardcoded text
+        } else {
+            post.excerpt
+        }
+        val isTimeIconVisible = postCardType == PostCardType.SCHEDULED && excerpt != null
+
+        PostItem(
+                title = post.title?.let { UiStringText(it) } ?: UiStringRes(R.string.my_site_untitled_post),
+                excerpt = excerpt?.let { UiStringText(it) },
+                featuredImageUrl = post.featuredImageUrl,
+                isTimeIconVisible = isTimeIconVisible
+        )
```

Comment:
Suggestion (💡): I was playing around with this, the `PostCardDraftOrScheduled` data class and its inner `PostItem` and something was itching me in terms of whether this can become a little more separated and/or simpler.

I then went and started splitting the `PostCardDraftOrScheduled` into a `DraftPostCard` and a `ScheduledPostCard`, which guide me through some refactorings. Long story short, I managed to end up with the following solution. Please consider that, along with it's pros and cons, but feel free to ignore it as well:

`PostItemType` (renamed from `PostCardType`)
```
enum class PostItemType(val id: Int) {
    DRAFT(0),
    SCHEDULED(1)
}
```

`PostCard`
```
        sealed class PostCard(
            open val title: UiString,
            open val postItems: List<PostItem>
        ) : Card(POST_CARD) {
            data class DraftPostCard(
                override val title: UiString,
                override val postItems: List<PostItem>
            ) : PostCard(title, postItems)

            data class ScheduledPostCard(
                override val title: UiString,
                override val postItems: List<PostItem>
            ) : PostCard(title, postItems)

            data class PostItem(
                val type: PostItemType,
                val title: UiString,
                val excerpt: UiString?,
                val featuredImageUrl: String?
            )
        }
```

`PostCardBuilder`
```
    private fun List<Post>.createDraftPostsCard() = DraftPostCard(
            title = UiStringRes(R.string.my_site_post_card_draft_title),
            postItems = mapToDraftPostItems()
    )

    private fun List<Post>.createScheduledPostsCard() = ScheduledPostCard(
            title = UiStringRes(R.string.my_site_post_card_scheduled_title),
            postItems = mapToScheduledPostItems()
    )

    private fun List<Post>.mapToDraftPostItems() = map { post ->
        PostCard.PostItem(
                type = PostItemType.DRAFT,
                title = post.title?.let { UiStringText(it) } ?: UiStringRes(R.string.my_site_untitled_post),
                excerpt = post.excerpt?.let { UiStringText(it) },
                featuredImageUrl = post.featuredImageUrl
        )
    }

    private fun List<Post>.mapToScheduledPostItems() = map { post ->
        PostCard.PostItem(
                type = PostItemType.SCHEDULED,
                title = post.title?.let { UiStringText(it) } ?: UiStringRes(R.string.my_site_untitled_post),
                excerpt = UiStringText("Today at 1:04 PM"),
                featuredImageUrl = post.featuredImageUrl
        )
    }
```

`PostCardViewHolder`
```
    fun bind(card: PostCard) = with(binding) {
        mySiteToolbar.update(card)
        (postItems.adapter as PostItemsAdapter).update(card.postItems)
    }

    private fun MySiteCardToolbarBinding.update(card: PostCard) {
        mySiteCardToolbarTitle.text = uiHelpers.getTextOfUiString(
                itemView.context,
                card.title
        )
    }
```

`PostItemViewHolder`
```
    fun bind(postItem: PostItem) = with(binding) {
        uiHelpers.setTextOrHide(title, postItem.title)
        uiHelpers.setTextOrHide(excerpt, postItem.excerpt)
        postItem.featuredImageUrl?.let { bindFeatureImage(it) }
        bindIconTime(postItem.type)
    }

    private fun PostItemBinding.bindFeatureImage(featuredImageUrl: String) {
        imageManager.loadImageWithCorners(
                featuredImage,
                PHOTO_ROUNDED_CORNERS,
                featuredImageUrl,
                uiHelpers.getPxOfUiDimen(
                        WordPress.getContext(),
                        UIDimenRes(dimen.my_site_post_item_image_corner_radius)
                )
        )
        featuredImage.setVisible(true)
    }

    private fun PostItemBinding.bindIconTime(type: PostItemType) {
        if (type == SCHEDULED) iconTime.setVisible(true)
    }
```

`post_item.xml`
```
    <ImageView
        android:id="@+id/featured_image"
        android:layout_width="@dimen/my_site_post_item_image_size"
        android:layout_height="@dimen/my_site_post_item_image_size"
        android:contentDescription="@string/featured_image_desc"
        android:scaleType="centerCrop"
        android:visibility="gone"
        app:layout_constraintBottom_toBottomOf="parent"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintTop_toTopOf="@id/title"
        tools:srcCompat="@tools:sample/avatars"
        tools:visibility="visible" />
```

PS.1: I renamed the enum to `PostItemType` because the `type` is now used within the `PostItem` data class and as such it felt that changing that was appropriate.
PS.2: I removed the `isFeaturedImageVisible` from the `PostItem` since it felt as being an unnecessary duplication. However, I am not sure if you are thinking of it from another perspective. Instead, I updated the `featured_image` image view to a default visibility of `gone`, just like you did with the `icon_time` image view. If we expect that most of the post will have a feature image, we might then want to have the default visibility to `visible`, and only set the `featuredImage.setVisible(false)` when that is not presented, as such this will become the opposite of the above solution.
PS.3: I am very unsure about why the `featuredImageCornerRadius` was part of the `PostItem` and thus I took the liberty of removing it and (kind of) hardcoding it within the `PostItemViewHolder` directly. Let me know if I am missing something.
PS.4: The `PostCardBuilderTest` will need to be updated as well, but nothing too dramatic here...

Reply:
Thank you @ParaskP7 for all the suggestions here and discussions in our call. 

As discussed, both the existing and suggested approaches are fine and each one has its pros and cons considering
- what do we consider as a post item (1:1 mapping with an existing post?)
- whether other cards need a post items list?
- deeper nesting in the `PostCard` sealed class
- code duplication in layouts (draft and scheduled cards have more-or-less same UI structure) vs nullability 

With all that in mind, we decided to revisit the approach after all the post cards are implemented and merged. 

> PS.1: I renamed the enum to PostItemType because the type is now used within the PostItem data class and as such it felt that changing that was appropriate.

As discussed, we will reconsider it later based on what approach we take up.

> PS.2: I removed the isFeaturedImageVisible from the PostItem since it felt as being an unnecessary duplication. However, I am not sure if you are thinking of it from another perspective. Instead, I updated the featured_image image view to a default visibility of gone, just like you did with the icon_time image view. If we expect that most of the post will have a feature image, we might then want to have the default visibility to visible, and only set the featuredImage.setVisible(false) when that is not presented, as such this will become the opposite of the above solution.

The field `isFeaturedImageVisible` was added mainly to avoid any logic in the VH (`postItem.featuredImageUrl != null` ) and to directly use it in the tests to know if the image is visible or not. To avoid duplication, I've accepted your suggestion in a6ea9bc.

> PS.3: I am very unsure about why the featuredImageCornerRadius was part of the PostItem and thus I took the liberty of removing it and (kind of) hardcoding it within the PostItemViewHolder directly. Let me know if I am missing something.

Suggestion accepted in a6ea9bc.

> PS.4: The PostCardBuilderTest will need to be updated as well, but nothing too dramatic here...

N/A

Reply:
👋 @ashiagr !

Thank you so much for this nice summary of what was discussed and for all the little improvements that you made based on my `PS`, much appreciated. 🙏